src/logic/player.cc

```diff
@@ -1418,39 +1417,45 @@ void Player::remove_seer(const MapObject& m, const Area<FCoords>& a) {
 }
 
 // Checks if any of our buildings or bobs, or one of one of our team mates, is seeing this node
-bool Player::should_see(const FCoords& f) const {
+bool Player::should_see(const FCoords& f, std::list<const MapObject*>& nearby_objects) const {
 	for (const PlayerNumber& p : team_player_) {
 		const Player& player = *egbase().get_player(p);
 		if (player.revealed_fields_.count(f)) {
 			return true;
 		}
-		for (const MapObject* mo : player.seers_) {
-			if (mo->descr().type() >= MapObjectType::BUILDING) {
-				upcast(const Building, b, mo);
-				assert(b);
-				if (b->is_seeing() &&
-				    egbase().map().calc_distance(f, b->get_position()) <= b->descr().vision_range()) {
-					return true;
-				}
-			} else {
-				// currently only buildings and bobs can see fields
-				upcast(const Bob, b, mo);
-				assert(b);
-				if (egbase().map().calc_distance(f, b->get_position()) <= b->descr().vision_range()) {
-					return true;
-				}
+	}
+	for (const MapObject* mo : nearby_objects) {
+		if (mo->descr().type() >= MapObjectType::BUILDING) {
+			assert(is_a(Building, mo));
+			const Building* b = static_cast<const Building*>(mo);
```

Comment:
How about keeping two lists of seers, one for `Building*` and one for `Bob*`, so we don't need to upcast all the time?

Reply:
For some reason I didn't want to do it but on second thought that should be worth it.

Reply:
Ah, I remember the reason. It would be tricky to have two lists without duplicating the code in multiple places. I thought a better solution would be to make a new type for seers. And I decided against it because I wanted to keep this diff (and the work involved) small as it's just a temporary solution that will be replaced later.